emcc.py

```diff
@@ -1128,14 +1128,29 @@ def get_last_setting_change(setting):
     # Set ASM_JS default here so that we can override it from the command line.
     shared.Settings.ASM_JS = 1 if shared.Settings.OPT_LEVEL > 0 else 2
 
+    # Remove the default _main function from shared.Settings.EXPORTED_FUNCTIONS.
+    # We do this before the user settings are applied so it effect the default value only and a
+    # user could use `--no-entry` and still export main too.
+    if options.no_entry:
+      shared.Settings.EXPORTED_FUNCTIONS.remove('_main')
+
     # Apply -s settings in newargs here (after optimization levels, so they can override them)
     apply_settings(settings_changes)
 
-    if options.no_entry and '_main' in shared.Settings.EXPORTED_FUNCTIONS:
-      shared.Settings.EXPORTED_FUNCTIONS.remove('_main')
-
     shared.verify_settings()
 
+    if options.no_entry or '_main' not in shared.Settings.EXPORTED_FUNCTIONS:
+      shared.Settings.EXPECT_MAIN = 0
+
+    if shared.Settings.STANDALONE_WASM:
+      # In STANDALONE_WASM mode we either build a command or a reactor.
+      # See https://github.com/WebAssembly/WASI/blob/master/design/application-abi.md
+      # For a command we always want EXIT_RUNTIME=1
+      # For a reactor we always want EXIT_RUNTIME=0
+      if 'EXIT_RUNTIME' in settings_changes:
+        exit_with_error('EXIT_RUNTIME not compatible with STANDALONE_WASM (use --no-entry to build a reactor)')
```

Comment:
instead of checking for `settings_changes`, how about if we apply the policy just mentioned if the user does not make a change. And then we can just compare the actual value on Settings to the proper value.

the advantage is that right now it seems like if the user doesn't set EXIT_RUNTIME then it has the right value for a reactor, but not for a command (as it's 0 by default), and so it's wrong according to the policy.

Reply:
My thought was that we don't want to entourage users of STANDALONE_WASM mode to specify EXIT_RUNTIME on the commandline at all.   We want them to completely remove this setting from the command line and let it be driven by the presence of `main` instead.   In standalone mode this setting becomes more of an internal setting I guess?

Reply:
I agree, but I still think it would be nice to have it be consistent internally. That is, for us to set it to the right value. Otherwise code will need to check not just `EXIT_RUNTIME` but also whether `STANDALONE_WASM`.

Reply:
I don't understand what you are asking for.  This code right does force EXIT_RUNTIME to be be internally consistent when using STANDALONE_WASM.     Its sets it unconditionally to the correct/consistent value on the line below this comment.  We only error out if the user tries to override this by also setting it redundantly on the command line.  Maybe I'm misunderstanding something?

Reply:
Sorry, I wasn't clear enough. Overall lgtm, but I think it might be nicer to not error if the user sets the value of `EXIT_RUNTIME` to the proper value anyhow. But as this is for standalone wasm, not many existing users, so I don't feel strongly.

But if we do things this way, how about making the error message here: `Picking EXIT_RUNTIME is not compatible with STANDALONE_WASM, where the value is determined automatically (--no-entry builds a reactor, which never exits the runtime, while a program always exits)` (is "program" the right word for a non-reactor?)

Reply:
Done